Skinned Mesh attributes

Squareys opened the pull request:
Hey @mosra !

here's some initial draft for Skinning stuff to start discussion.

Cheers,
Jonathan

**Open Questions**
 1.Naming of Attributes

**TODOs**
 - [x] Add `Weights` and `JointIds` to AttributeNames and Generic attributes.
 - [x] MeshData & Testing
 - [ ] Documentation?

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -358,7 +358,26 @@ template<UnsignedInt dimensions> struct Generic {
+    // TODO: E.g. gltf supports importing multiple sets of 4 indices (JOINTS_0 [0;3], JOINTS_1 [4;7])
+    // I suppose these would require special shaders and it's fine to only support 4 for the GenericShaders
+    // TODO: This is called "Joints" in gltf.
+    typedef GL::Attribute<7, Vector4> JointIndices;
```
```suggestion
    typedef GL::Attribute<7, Vector4> JointIds;
```

Also isn't this an integer attribute, when it's a ID? All other attribs are singular, so maybe make this `JointId` and `Weight` also? Kinda on the fence about this tho, because there's `TextureCoordinates` also.

Squareys replied:
Since it's always four at a time (i.e. four joints that might influence a vertex at any given time), I think conveying it's "more than one" through the name would make sence in my oppinion.

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -358,7 +358,26 @@ template<UnsignedInt dimensions> struct Generic {
+     * Corresponds to @ref Trade::MeshAttribute::Weights.
+     */
+    // TODO: E.g. gltf supports importing multiple sets of 4 weights (WEIGHTS_0 [0;3], WEIGHTS_1 [4;7])
+    // I suppose these would require special shaders and it's fine to only support 4 for the GenericShaders
```
Yep, and that's why I reserved the "2nd vertex weights" etc above. This would be then a `SecondaryWeights` attribute, and yes the shader needs to explicitly support that (it's not fine to hard-cap the support to just 4, no :P).

Hmm, now i'm thinking if it's actually possible for the shader to access both primary and secondary weights through the same accessor (maybe through a `mat2x4` where the first column is bound to id 6 and the second to id 10?)

Squareys replied:
Why would it need to access both weight sets through the same accessor? The Gltf 2.0 overview seems to use a very "unrolled" way of multiplying weights with bone transforms, no matrix multiplications there:

```glsl

uniform mat4 u_jointMatrix[12];
in vec4 position;
in vec4 jointIds;
in vec4 weights;

void main(void) {
    mat4 skinMatrix =
        weights.x*jointMatrix[int(jointIds.x)] +
        weights.y*jointMatrix[int(jointIds.x)] +
        weights.z*jointMatrix[int(jointIds.x)] +
        weights.w*jointMatrix[int(jointIds.x)];

    gl_Position = modelViewProjection*skinMatrix*position;
}
```

mosra replied:
Shouldn't that be `jointIds.xyzw` also? I mean the case with

```glsl
in vec4 jointIds;
in vec4 secondaryJointIds;
in vec4 weights;
in vec4 secondaryWeights;
```

is kinda ugly, so I'm wondering if this could be turned into some

```glsl
in float jointIds[8];
in float weights[8];
```

instead, with the `8` being possibly passed in during compilation (so e.g. we can support 5 influences without always doing calculations for 8.

Squareys replied:
> Shouldn't that be jointIds.xyzw also?

Yes! => fixed

> pseudocode

This will lead to some ugly macro stuff if you do it this way. Instead with the duplicate attributes, it'd be just a block like `#ifdef MORE_BONES` which would be added rather than replacing the code that works for 4 joints/2 attributes.

mosra replied:
> ugly macro

why? this is what I wanted to achieve

```glsl
mat4 skinMatrix;
for(int i = 0; i != HOW_MANY_THEM_BONES; ++i)
    skinMatrix += weights[i]*jointMatrices[int(jointIds[i])];
```

Squareys replied:
Wouldn't `mat2x4 jointIds` return a vec4 on `jointIds[0]` though?

mosra replied:
This is becoming endless :sweat_smile: 

Yes it would. I'm referring to the `in float jointIds[8];` above.

mosra commented on src/Magnum/Trade/MeshData.h:
```diff
@@ -1717,6 +1736,54 @@ class MAGNUM_TRADE_EXPORT MeshData {
+         * @see @ref jointIndicesInto(), @ref attributeFormat(),
+         *      @ref isVertexFormatImplementationSpecific()
+         */
+        Containers::Array<Vector4us> jointIndicesAsArray(UnsignedInt id = 0) const;
```
I think should always be the largest type (even if it will never be as big):

```suggestion
        Containers::Array<Vector4ui> jointIndicesAsArray(UnsignedInt id = 0) const;
```

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -483,7 +497,8 @@ template<> struct Generic<3>: BaseGeneric {
     typedef GL::Attribute<5, Vector3> Normal;
-    /* 6, 7 reserved for vertex weights / bone IDs */
+    typedef GL::Attribute<6, Vector4> Weights;
+    typedef GL::Attribute<7, Vector4> JointIds;
```
```suggestion
    typedef GL::Attribute<7, Vector4ui> JointIds;
```

Also, where's the 2D variant?

Squareys replied:
Why wouldn't this work for 2D?

mosra replied:
Because it's defined only in the 3D specialization of the class ;)

mosra commented on src/Magnum/Trade/MeshData.h:
```diff
@@ -147,6 +147,23 @@ enum class MeshAttribute: UnsignedShort {
+    /**
+     * Joint IDs. Type is usually @ref VertexFormat::Vector4ui, but can be also
+     * @ref VertexFormat::Vector4us or @ref VertexFormat::Vector4ub.
+     * @ref Shaders::Generic::JointIds.
```
```suggestion
     * Corresponds to @ref Shaders::Generic::JointIds.
```

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -150,13 +150,13 @@ both bitangents and object ID for instancing, \n
 <td>10</td>
 <td colspan="2">
-* *Reserved* --- 2nd vertex weights
+@ref Weights
```
This is still wrong -- either add `SecondaryWeights`, `SecondaryJointIds` or revert this and the below change.

Squareys replied:
Will make it SecondaryWeights 👍

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -126,13 +126,13 @@ both bitangents and object ID for instancing, \n
 <td>7</td>
 <td colspan="3">
-* *Reserved* --- bone indices
+@ref JointIndices
```
```suggestion
@ref JointIds
```

mosra commented:
What's left for this to be useful is adding those to `MeshTools::compile()` -- but to test it there, you would need some shader that supports them, and at that point you have implemented skinning completely :)

Squareys commented:
Sure, it looks straight forward enough to do so, and I'll have to eventually anyways, so why not rip off the bandage.

Squareys commented:
@mosra I started implementing Skinning in Phong to test the additions to compile(), but this would blow up the PR significantly. Could we split Phong shader implementation, compile() and test off into a separate PR?

mosra commented:
Yup, good idea -- the Phong implementation also needs tests on its own, so it makes sense to do all that separately.

Squareys commented:
Split Phong implementation into Squareys:phong-skinning, will create the PR for that once this is merged.
Rebased this and squashed some commits, I think this is ready for review :)

mosra commented on src/Magnum/Trade/Test/MeshDataTest.cpp:
```diff
@@ -2500,6 +2522,90 @@ void MeshDataTest::objectIdsIntoArrayInvalidSize() {
+    CORRADE_COMPARE(out.str(),
+        "Trade::MeshData::jointIdsInto(): expected a view with 3 elements but got 2\n");
+}
+
```
The `attributeNotFound()` and `implementationSpecificVertexFormatWrongAccess()` tests need to be extended to check asserts in the new accessors.

mosra commented on src/Magnum/Trade/MeshData.h:
```diff
@@ -147,6 +147,23 @@ enum class MeshAttribute: UnsignedShort {
+     * Corresponds to @ref Shaders::Generic::JointIds.
+     * @see @ref MeshData::jointIdsAsArray()
+     */
+    JointIds,
```
I'm wondering -- would it make sense to support also one/two/three-component weights and joints? I guess it would be useful for file size savings. Can you check if there are example glTF files with less than 4 weights?

This won't complicate the implementation too much, a similar thing is already done for colors for example.

Squareys replied:
The accessor is always a 4-component vector, so in case there is a file like this, it would have 4 but the unused components would be zero weight

mosra replied:
In (unextended) glTF yes, but in general not. I don't see why glTF couldn't support `SCALAR`, `VEC2` and `VEC3` also.

mosra replied:
It would be just JointIds always, but allowing also one/two/three-component types instead of just Vectot4ui.

`MeshData::jointIdsInto()` will then need to zero-fill the unused components like `colorsInto()` does for three-component colors or `positions3DInto()` does for two-component positions, yes.

mosra commented on src/Magnum/Shaders/Generic.h:
```diff
@@ -478,6 +492,11 @@ struct BaseGeneric {
+    typedef GL::Attribute<7, Vector4> JointIds;
+
+    typedef GL::Attribute<10, Vector4> SecondaryWeights;
+    typedef GL::Attribute<11, Vector4> SecondaryJointIds;
```
Seeing #444, this PR should contain also the `generic.glsl` changes and corresponding additions to `GenericTest::glslMatch()`

mosra commented:
Merged with various changes as 7f94a6f914c834e0d693cf2a8d62eb1d40492699 and f447e994d271c8688da3cf7f04a1a2643f42cc21. The primary difference is that `MeshAttribute::JointIds` and `Weights` is now an array attribute in order to allow a single attribute to have any number of components -- two, four but also for example seven. This allows both for better packing in memory (compared to always having 4 or 8 components) as well as easier processing because with > 4 components the attribute isn't split into two. On the other hand it's still possible to have two `JointIds` and two `Weights` for example, and have the second set in a different (e.g., less precise) type.

The builtin shader bindings slots stay four component, with the intention that (a) the shader is either compiled for an appropriate per-vertex component count, (b) a dynamic per-vertex component count is passed to it, or (c) the mesh is repacked to always contain the maximum number of components the shader expects. 

A major change with the builtin attribute is that they have a flipped order compared to this PR -- `JointIds` and `SecondaryJointIds` are 6 and 10 instead of 7 and 11, `Weights` and `SecondaryWeights` are 7 and 11 instead of 6 and 10. This order made more sense to me, i.e. "joint i has a weight j" instead of "weight i is applied to joint j".
